### PR TITLE
[Ansible] Etcd automatic cert generation

### PR DESCRIPTION
**Pull request by galexrt**
I played around with automatic certificate generation for etcd.

~~I haven't had time to test this yet~~, but I can already say that switching an etcd cluster from http to https most likely would require some manual work (switching the etcd member urls to https).

I would love to get some feedback on my changes for adding automatic etcd cert generation and hopefully see this merged to make securing etcd simpler using the Ansible deployment simpler.

Currently there aren't that many configurable variables added for the certificate generation, but this could be also implemented later on.

### CODE REVIEW

**Comment by ingvagabund** [comment]
@jasonbrooks Jason, do you know someone who could help with this?

**Comment by galexrt** [comment]
I just added etcd client certificates configuration for Flanneld.

**Comment by jasonbrooks** [comment]
I gave this a test with fedora atomic. I needed to make some changes to get things running w/ atomic:

```
diff --git a/ansible/roles/etcd/tasks/gen-certs.yml b/ansible/roles/etcd/tasks/gen-certs.yml
index c510472..068a43a 100644
--- a/ansible/roles/etcd/tasks/gen-certs.yml
+++ b/ansible/roles/etcd/tasks/gen-certs.yml
@@ -12,14 +12,6 @@
   set_fact:
     etcd_machine_address: "{{ etcd_machine_address }}"
 
-- name: Overwrite make-ca-cert.sh from local copy
-  copy:
-    src: make-ca-cert.sh
-    dest: "{{ etcd_script_dir }}"
-    mode: "0500"
-  changed_when: false
-  delegate_to: "{{ groups['etcd'][0] }}"
-
 - name: Create etcd script directory
   file:
     path: "{{ etcd_script_dir }}"
@@ -29,6 +21,14 @@
   delegate_to: "{{ groups['etcd'][0] }}"
   run_once: true
 
+- name: Overwrite make-ca-cert.sh from local copy
+  copy:
+    src: make-ca-cert.sh
+    dest: "{{ etcd_script_dir }}"
+    mode: "0500"
+  changed_when: false
+  delegate_to: "{{ groups['etcd'][0] }}"
+
 - name: Run create cert script on first etcd server
   command:
     "{{ etcd_script_dir }}/make-ca-cert.sh"
diff --git a/ansible/roles/etcd/tasks/main.yml b/ansible/roles/etcd/tasks/main.yml
index 9f9fc56..85dbfde 100644
--- a/ansible/roles/etcd/tasks/main.yml
+++ b/ansible/roles/etcd/tasks/main.yml
@@ -1,4 +1,7 @@
 ---
+- name: Override etcd facts
+  include: override_facts.yml
+
 - name: Install etcd
   include: install.yml
   tags: install
```

After that, though, the "Run create cert script on first etcd server" task dumped a ton of binary-looking output to my terminal before failing -- I haven't been able to make sense of it so far.

**Comment by galexrt** [comment]
I'll take a look at the issue and add the atomic fix.

**Comment by galexrt** [comment]
@jasonbrooks The vagrant setup isn't working. The VirtualBox VMs just won't up eth1 and get an IP, and Fedora 23 image isn't available anymore (seems that Fedora 25 is only available now; switching to Fedora 25 worked, but as written the eth1 just wouldn't come up and get an IP).
I tested the improvements/changes I now added directly.

**Comment by jasonbrooks** [comment]
@galexrt I sent a PR for the fedora vagrant box bit yesterday: #2388 

I didn't test this PR w/ vagrant -- I tested w/ VMs running fedora atomic 25 and regular f25 and got the same weird output result. My VMs have one nic -- this doesn't somehow rely on multiple nics, does it?

**Comment by galexrt** [comment]
@jasonbrooks I just pushed new changes two minutes ago. That address the issues you mentioned.

**Comment by jasonbrooks** [comment]
@galexrt Also, I imagine that coreos might have a similar issue w/ the scripts dir. I copied the override for the etcd facts from this one for the kubernetes role, and there's a different coreos scripts dir location: https://github.com/kubernetes/contrib/blob/master/ansible/roles/kubernetes/tasks/override_facts.yml

**Comment by galexrt** [comment]
@jasonbrooks I added it in my latest push.

***

To your last question:
My changes don't need a second (or more) nics, but the Vagrantfile creates `eth1` and tells Ansible to use it as  the default interface (see https://github.com/kubernetes/contrib/blob/master/ansible/vagrant/Vagrantfile#L209)

File: ansible/roles/kubernetes/tasks/override_facts.yml [hunk 1]
```diff
@@ -2,6 +2,7 @@
 - name: Update {{ kube_script_dir }} if this is atomic
   set_fact:
     kube_script_dir: "/usr/local/libexec/kubernetes"
+    etcd_script_dir: "/usr/local/libexec/etcd"
```
Comment (jasonbrooks) [inline ansible/roles/kubernetes/tasks/override_facts.yml]: This isn't working, because the the etcd task is executing before the kubernetes task does.

Comment (galexrt) [inline ansible/roles/kubernetes/tasks/override_facts.yml]: Fixed. I created a etcd `override_facts.yml`.

File: ansible/roles/etcd/tasks/override_facts.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+---
+- name: Update {{ etcd_script_dir }} if this is atomic
+  set_fact:
+    etcd_script_dir: "/usr/local/libexec/etcd"
+  when: is_atomic and kube_script_dir == "/usr/libexec/etcd"
```
Comment (jasonbrooks) [inline ansible/roles/etcd/tasks/override_facts.yml]: needs to be etcd_script_dir

Comment (galexrt) [inline ansible/roles/etcd/tasks/override_facts.yml]: Fixed.

**Comment by jasonbrooks** [comment]
I'm getting fails on this flannel task:

```
TASK [flannel : Make sure etcd certificate vars are set] ***********************
fatal: [fah-2.osas.lab]: FAILED! => {"changed": false, "failed": true, "msg": "etcd certificates weren't slurped. Aborting ..."}
skipping: [fah-1.osas.lab]
fatal: [fah-3.osas.lab]: FAILED! => {"changed": false, "failed": true, "msg": "etcd certificates weren't slurped. Aborting ..."}
```

**Comment by galexrt** [comment]
Mhh. In which inventory group are the two failing nodes?

**Comment by jasonbrooks** [comment]
They're the "nodes"

**Comment by galexrt** [comment]
@jasonbrooks I forgot to add the `etcd-certs` role to the playbooks. Please retest with the latest changes again.

**Comment by jasonbrooks** [comment]
@galexrt Yes, it's working for me now

**Comment by ingvagabund** [comment]
@redhat-k8s-bot test please

**Comment by ingvagabund** [comment]
``Boilerplate header is wrong for: /home/travis/gopath/src/k8s.io/contrib/ansible/roles/etcd/files/make-ca-cert.sh``

**Comment by galexrt** [comment]
@ingvagabund What is "Boilerplate header is wrong for ..." error? Is it because of the license header?

**Comment by ingvagabund** [comment]
```sh
$ ./hack/boilerplate/boilerplate.py -v ansible/roles/etcd/files/make-ca-cert.sh
Header in /home/jchaloup/Projects/src/github.com/kubernetes/contrib/ansible/roles/etcd/files/make-ca-cert.sh does not match reference, diff:

--- reference
+++ /home/jchaloup/Projects/src/github.com/kubernetes/contrib/ansible/roles/etcd/files/make-ca-cert.sh
@@ -1,4 +1,4 @@
-# Copyright YEAR The Kubernetes Authors.
+# Copyright 2017 The Kubernetes Authors.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
 # you may not use this file except in compliance with the License.

/home/jchaloup/Projects/src/github.com/kubernetes/contrib/ansible/roles/etcd/files/make-ca-cert.sh
```

It is because we have 2017 and the script allows only existence of <=2016.

**Comment by galexrt** [comment]
So should I change it to 2016 then?

**Comment by ingvagabund** [comment]
The Centos test can be ignored again.

> So should I change it to 2016 then?

I will fix the boilerplate script, minute.

**Comment by ingvagabund** [comment]
The fix is here https://github.com/kubernetes/contrib/pull/2395. Though, I can not approve the PR. Hope it will get merged fast cause this will block PR of anyone who creates new *.sh or *.go files.

**Comment by ingvagabund** [comment]
@galexrt feel free to re-run the tests again (e.g. ``git commit --amend && git push``). #2395 is merged

**Comment by galexrt** [comment]
@ingvagabund Test was green before I rebased. Please test again.

**Comment by ingvagabund** [comment]
@redhat-k8s-bot test please

**Pull request merged**